File: Days.33/C++/5.Convert-Normal-BST-into-Balanced-BST.cpp
// Convert a Normal BST ----> Balanced BST

#include<iostream>
#include<vector>
#include<queue>
#include<string>
using namespace std;

class Node
{
    public:
    int val;
    Node *left;
    Node *right;
    Node(int val)
    {
        this->val=val;
        this->left=this->right=nullptr;
    }
};


void Inputs(string &str,vector<string> &nodes)
{
    int i=0;
    string temp="";
    while(i<str.size())
    {
        if(str[i]==' ')
        {
            nodes.push_back(temp);
            temp="";
        }
        else
        {
            temp+=str[i];
        }
        i++;
    }
    nodes.push_back(temp);
}

Node* BuildTree(vector<string> &nodes)
{
    int n=nodes.size();
    if (n==0 or nodes[0]=="N")
    {
        return nullptr;
    }
    Node *root=new Node(stoi(nodes[0]));
    int i=1;
    queue<Node*> q;
    q.push(root);
    Node *curr;
    while (!q.empty() and i<n)
    {
        curr=q.front();
        q.pop();
        if(nodes[i]!="N")
        {
            curr->left=new Node(stoi(nodes[i]));
            q.push(curr->left);
        }
        i++;
        if(i>=n)
        {
            break;
        }
        if(nodes[i]!="N")
        {
            curr->right=new Node(stoi(nodes[i]));
            q.push(curr->right);
        }
        i++;
    }
    return root;
}

void Inorder(Node* root,vector<int> &inorder)
{
    if(root==nullptr)
    {
        return;
    }
    Inorder(root->left,inorder);
    inorder.push_back(root->val);
    Inorder(root->right,inorder);
}

int isBalanced(Node *root,bool &flag)
{
    if(root==nullptr)
    {
        return 0;
    }
    int left=isBalanced(root->left,flag);
    int right=isBalanced(root->right,flag);
    if (abs(left-right)>1)
    {
        flag=false;
    }
    return max(left,right)+1;
}

Node* BalancedBST(int start,int end,vector<int> &inorder)
{
    if(start>end)
    {
        return nullptr;
    }
    if(start==end)
    {
        return new Node(inorder[start]);
    }
    int mid=(start+(end-start)/2);
    Node *root=new Node(inorder[mid]);
    root->left=BalancedBST(start,mid-1,inorder);
    root->right=BalancedBST(mid+1,end,inorder);
    return root;
}

int main()
{
    string str;
    getline(cin,str);
    vector<string> nodes;
    Inputs(str,nodes);
    Node *root=BuildTree(nodes);
    cout<<"Normal BST Balanced or Not:\t";
    bool flag=true;
    isBalanced(root,flag);
    if(flag)
    {
        cout<<"Yes\n";
    }
    else{
        cout<<"No\n";
    }
    vector<int> inorder;
    Inorder(root,inorder);
    Node *newroot=BalancedBST(0,inorder.size()-1,inorder);
    cout<<"Balanced BST Balanced or Not\t";
    flag=true;
    isBalanced(newroot,flag);
    if(flag)
    {
        cout<<"Yes\n";
    }
    else{
        cout<<"No\n";
    }
    return 0;
}

File: Days.33/Python/5.Convert-Normal-BST-into-Balanced-BST.py
# Convert a Normal BST ----> Balanced BST

from collections import deque

class Node:
    def __init__(self,val):
        self.val=val
        self.left=None
        self.right=None

def BuildTree(str):
    if len(str)==0 or str[0]=='N':
        return None
    dq=deque()
    root=Node(int(str[0]))
    dq.append(root)
    i=1
    n=len(str)
    while dq and i<n:
        curr=dq.popleft()
        if str[i]!='N':
            curr.left=Node(int(str[i]))
            dq.append(curr.left)
        i+=1
        if i>=n:
            break
        if str[i]!='N':
            curr.right=Node(int(str[i]))
            dq.append(curr.right)
        i+=1
    return root

def Inorder(root,inorder):
    if root==None:
        return
    Inorder(root.left,inorder)
    inorder.append(root.val)
    Inorder(root.right,inorder)

def BST(start,end,inorder):
    if start>end:
        return None
    if start==end:
        return Node(inorder[start])
    mid=(start+end)//2
    root=Node(inorder[mid])
    root.left=BST(start,mid-1,inorder)
    root.right=BST(mid+1,end,inorder)
    return root

def BalancedBST(root):
    inorder=[]
    Inorder(root,inorder)
    return BST(0,len(inorder)-1,inorder)

def Balanced(root,flag):
    if root==None:
        return 0
    left=Balanced(root.left,flag)
    right=Balanced(root.right,flag)
    h=abs(left-right)
    if h>1:
        flag[0]=False
    return max(left,right)+1
def main():
    str=input().split()
    root=BuildTree(str)
    flag=[True]
    newroot=BalancedBST(root)
    Balanced(newroot,flag)
    if flag[0]==True:
        print('Tree is Balanced Now')
    else:
        print('Tree is Not Balanced')
main()

